### ISSUE TITLE
Different results with “xgboost” official package vs. xgboost from “caret” package in R

### ISSUE THREAD

**Issue by PPStef** [dmlc/xgboost#1371]
I am new to R programming language and I need to run "xgboost" for some experiments. The problem is that I need to cross-validate the model and get the accuracy and I found two ways that give me different results:

With "caret" using:

    library(mlbench)
    library(caret)
    library(caretEnsemble)
    dtrain <- read.csv("student-mat.csv", header=TRUE, sep=";")
    formula <- G3~.
    dtrain$G3<-as.factor(dtrain$G3)
    control <- trainControl(method="cv", number=10)
    seed <- 10
    metric <- "Accuracy"    
    fit.xgb <- train(formula, data=dtrain, method="xgbTree", metric=metric, trControl=control, nthread =4)
    fit.xgb
    fit.xgbl <- train(formula, data=dtrain, method="xgbLinear", metric=metric, trControl=control, nthread =4)
    fit.xgbl

Using the "xgboost" package and the following code:

     library(xgboost)
     printArray <- function(label, array){
     cat(paste(label, paste(array, collapse = ", "), sep = ": \n"), "\n\n")
     setwd("D:\\datasets")
     dtrain <- read.csv("moodle7original(AtributosyNotaNumericos).csv",        header=TRUE, sep=",")
     label <- as.numeric(dtrain[[11]])
     data <- as.matrix(sapply(dtrain, as.numeric))

    croosvalid <-
    xgb.cv(
    data = data,
    nfold = 10,
    nround = 10,
    label = label,
    prediction = TRUE,
    objective = "multi:softmax",
    num_class = 33
    )
    
    print(croosvalid)  
    printArray("Actual classes", label[label != croosvalid\$pred])  
    printArray("Predicted classes", croosvalid\$pred[label != croosvalid\$pred])  
    correctlyClassified <- length(label[label == croosvalid\$pred])  
    incorrectlyClassified <- length(label[label != croosvalid\$pred])  
    accurancy <- correctlyClassified * 100 / (correctlyClassified + incorrectlyClassified)  
    print(paste("Accurancy: ", accurancy))  

But the results differ very much on the same dataset. I usually get 99% accuracy on student performance dataset with the second snip of code and ~63% with the first one...  
I set the same seed on both of them.

Am I wrong with the second? Please tell me why if so!

**Comment by Laurae2** [comment]
Using caret, the parameters are not well tuned and here are the parameters that are likely to be used with the default length of tuning of 3 against your xgb.cv parameters:

Parameter | Caret | XGBoost
----------- | ------- | ----------
Depth | (1, 2, 3) | 6
Rounds | (50, 100, 150) | 10
Learning rate | (0.3, 0.4) | 0.3
Colsample_bytree | (0.6, 0.8) | 1

You can clearly notice the hyperparameters are not the same values, and if you reach 99% accuracy with only 10 rounds and a depth of 6, caret is likely to not have the same results using such grid search it is using (36 models, but with very bad grid search according to your xgb.cv against caret results). Therefore, it should be perfectly normal to have 63% for caret, especially if you have 30+ classes and using only a maximum depth of 3.
